src/Platforms/SYCL/SYCLDeviceManager.cpp

```diff
@@ -0,0 +1,171 @@
+//
+// File created by: Thomas Applencourt, apl@anl.gov, Argonne National Laboratory
+//
+//////////////////////////////////////////////////////////////////////////////////////
+
+
+#include "SYCLDeviceManager.h"
+#include <stdexcept>
+#include <string>
+#include <algorithm>
+#include <CL/sycl/backend.hpp>
+#include <level_zero/ze_api.h>
+#include <CL/sycl/backend/level_zero.hpp>
+#include "OutputManager.h"
+#include "determineDefaultDeviceNum.h"
+#if defined(_OPENMP)
+#include <omp.h>
+#endif
+
+namespace qmcplusplus
+{
+
+SYCLDeviceManager::SYCLDeviceManager(int& default_device_num, int& num_devices, int local_rank, int local_size)
+    : sycl_default_device_num(-1)
+{
+  // Potentially multiple GPU platform.
+  std::vector<sycl::platform> platforms = sycl::platform::get_platforms();
+  if (platforms.empty())
+    throw std::runtime_error("Cannot find SYCL platforms!");
+
+    // find out devices from the first platform with GPUs.
+#if defined(ENABLE_OFFLOAD)
+  visible_devices = xomp_get_sycl_devices();
+#else
+  std::vector<sycl::device> devices;
+  app_log() << "Visible SYCL platforms are :" << std::endl;
+  for (auto& platform : platforms)
+  {
+    std::vector<sycl::device> gpu_devices = platform.get_devices(sycl::info::device_type::gpu);
+    const auto gpu_count                  = gpu_devices.size();
+    bool selected                         = false;
+    if (devices.empty() && gpu_count > 0)
+    {
+      selected = true;
+      devices  = std::move(gpu_devices);
+    }
+    app_log() << (selected ? " ** " : "    ") << platform.get_info<sycl::info::platform::name>() << " with "
+              << gpu_count << " GPUs." << std::endl;
+  }
+  app_log() << std::endl;
+
+  visible_devices.reserve(devices.size());
+  for (int id = 0; id < devices.size(); id++)
+    visible_devices.push_back({sycl::context(devices[id]), devices[id]});
+#endif
+
+  const size_t sycl_device_count = visible_devices.size();
+  if (num_devices == 0)
+    num_devices = sycl_device_count;
+  else if (num_devices != sycl_device_count)
+    throw std::runtime_error("Inconsistent number of SYCL devices with the previous record!");
+  if (sycl_device_count > local_size)
+    app_warning() << "More SYCL devices than the number of MPI ranks. "
+                  << "Some devices will be left idle.\n"
+                  << "There is potential performance issue with the GPU affinity.\n";
+  if (num_devices > 0)
+  {
+    sycl_default_device_num = determineDefaultDeviceNum(sycl_device_count, local_rank, local_size);
+    if (default_device_num < 0)
+      default_device_num = sycl_default_device_num;
+    else if (default_device_num != sycl_default_device_num)
+      throw std::runtime_error("Inconsistent assigned SYCL devices with the previous record!");
+
+    default_device_queue =
+        sycl::queue(visible_devices[sycl_default_device_num].context, visible_devices[sycl_default_device_num].device);
+  }
+}
+
+#if defined(_OPENMP)
```

Comment:
This would be the only code the we can object to here.  What is it needed for other than the SYCLDeviceManager constructor?

Reply:
CUDA has a primary context notion while SYCL requires manage context/devices explicitly. Thus we need OpenMP interop to make sure a unique context is shared between SYCL and OpenMP parts.